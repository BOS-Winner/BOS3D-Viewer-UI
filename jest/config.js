module.exports = {
  rootDir: '../',
  testMatch: [
    "**/__tests__/**/*.test.{js,jsx}",
  ],
  moduleNameMapper: {
    "\\.less$": "identity-obj-proxy",
    "\\.css$": "<rootDir>/jest/__mocks__/styleMock.js",
    "Libs(.*)$": "<rootDir>/src/Libs$1",
    "UIutils(.*)$": "<rootDir>/src/UI/utils$1",
    "^Base(.*)$": "<rootDir>/src/UI/Base$1",
    "^BOS3D(.*)$": "<rootDir>/jest/__mocks__/BOS3D$1",
    "^BOS2D$": "<rootDir>/jest/__mocks__/BOS2D.js",
    "^BOS3DUI$": "<rootDir>/jest/__mocks__/BOS3DUI.js",
    "^jest/(.*)$": "<rootDir>/jest/$1",
    "^mock/(.*)$": "<rootDir>/jest/__mocks__/$1",
  },
  moduleFileExtensions: ["js", "jsx"],
  moduleDirectories: ["node_modules"],
  transform: {
    "\\.(js|jsx)": "babel-jest",
    "\\.(jpg|jpeg|png|gif|svg)$": "<rootDir>/jest/__mocks__/fileMock.js",
  },
  transformIgnorePatterns: [
    // Change MODULE_NAME_HERE to your module that isn't being compiled
    // "/node_modules/(?!three|lodash-es).+\\.js$"
    "/node_modules/core-js",
    "/node_modules/@testing-library",
    "/node_modules/@babel",
  ],
  coveragePathIgnorePatterns: [
    "/node_modules/",
    "/jest/"
  ],
};
