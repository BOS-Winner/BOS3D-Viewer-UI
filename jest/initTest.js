import EventEmitter from "events";
import { createStore } from "redux";
import BOS3D from 'BOS3D';
import reducer from "../src/UI/reducer";

export function genStore() {
  const ee = new EventEmitter();
  const viewer3D = new BOS3D.Viewer();

  return createStore(reducer, {
    system: {
      viewer3D,
      BIMWINNER: { BOS3D },
      apiVersion: 'api',
      eventEmitter: ee,
    }
  });
}
