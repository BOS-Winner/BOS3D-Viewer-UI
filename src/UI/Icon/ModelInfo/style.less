.model-info {
  margin: 14px 0 0 14px;

  .item {
    display: flex;
    // margin-top: 8px;
    width: 100%;
    justify-content: space-around;
    align-items: center;
    .name {
      width: 80px;
      margin-top: 12px;
    }
  }
}
