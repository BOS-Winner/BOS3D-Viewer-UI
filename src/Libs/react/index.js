import React from "react";
import useUpdateEffect from "./useUpdateEffect";

React.useUpdateEffect = useUpdateEffect;

export default React;
// export { default as useUpdateEffect } from "./useUpdateEffect";
