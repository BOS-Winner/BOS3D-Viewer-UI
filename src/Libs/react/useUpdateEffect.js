import { useEffect, useRef } from "react";

export default function (effect, deps) {
  const isFirstRender = useRef(true);
  useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
    } else {
      effect();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);
}
