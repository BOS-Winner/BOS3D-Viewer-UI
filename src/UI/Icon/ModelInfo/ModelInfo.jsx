import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Icon from "Base/Icon";
import Modal from "Base/Modal";
import { Statistic } from 'antd';
import style from "./style.less";
// import iconImg from "../img/white/modelInfo.png";
import { AntdIcon } from '../../utils/utils';
import iconstyle from '../../Toolbar/bottom.less';
import { DEFAULT_MODAL_PLACE } from "../../constant";

class ModelInfo extends React.Component {
  static propTypes = {
    viewer: PropTypes.object.isRequired,
    mode: PropTypes.string.isRequired,
  }

  constructor(props) {
    super(props);
    this.state = {
      visible: false,
      trianglesCount: 0,
      componentCount: 0,
    };
  }

  componentDidUpdate(prevProps, prevState) {
    if (!prevState.visible && this.state.visible) {
      const info = this.props.viewer.getViewerImpl().modelManager.getStatisticsInfo();
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        trianglesCount: info.trianglesCount,
        componentCount: info.componentCount,
      });
    }
    if (prevProps.mode !== this.props.mode && this.props.mode === "漫游模式") {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        visible: false,
      });
    }
  }

  switchVisible(visible) {
    this.setState({
      visible,
    });
  }

  render() {
    const { visible, trianglesCount, componentCount } = this.state;

    return (
      <div
        title="模型信息"
        role="button"
        tabIndex={0}
        onClick={() => this.switchVisible(!visible)}
      >
        <Icon
          icon={<AntdIcon type="iconmodelinformation" className={iconstyle.icon} />}
          title=""
          selected={visible}
          showTitle={false}
        />
        <Modal
          title="模型信息"
          visible={visible}
          top={DEFAULT_MODAL_PLACE.modelinfo.top}
          right={DEFAULT_MODAL_PLACE.modelinfo.right}
          left={DEFAULT_MODAL_PLACE.modelinfo.left}
          width="350px"
          height="160px"
          minWidth={300}
          minHeight={160}
          onCancel={() => this.switchVisible(false)}
          viewportDiv={this.props.viewer.viewportDiv}
        >
          <div className={style.modelInfo}>
            <div className={style.item}>
              <div className={style.name}>三角面数：</div>
              {/* {trianglesCount} */}
              <Statistic value={trianglesCount} title={null} valueStyle={{ color: "#fff" }} />
            </div>
            <div className={style.item}>
              <div className={style.name}>构件数：</div>
              {/* {componentCount} */}
              <Statistic value={componentCount} title={null} valueStyle={{ color: "#fff" }} />
            </div>
          </div>
        </Modal>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  viewer: state.system.viewer3D,
  mode: state.button.mode,
});
const mapDispatchToProps = () => ({});
const WrappedContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(ModelInfo);
export default WrappedContainer;
