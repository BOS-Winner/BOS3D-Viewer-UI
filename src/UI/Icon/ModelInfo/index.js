import ModelInfo from "./ModelInfo";

export default ModelInfo;
