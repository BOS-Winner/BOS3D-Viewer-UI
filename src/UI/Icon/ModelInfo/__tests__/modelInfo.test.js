import React from 'react';
import { render, act } from '@testing-library/react';
import { Provider } from 'react-redux';
import { genStore } from 'jest/initTest';
import MockModal from "mock/Modal";
import ModelInfo from "../index";

const store = genStore();
jest.mock('Base/Modal', () => props => <MockModal {...props} />);
let gContainer;
let gabTitle;

describe("modelInfo", () => {
  beforeEach(() => {
    const { container, getAllByTitle } = render((
      <Provider store={store}>
        <ModelInfo />
      </Provider>
    ));
    gContainer = container;
    gabTitle = getAllByTitle;
  });

  it("should render dom", () => {
    expect(gContainer).toMatchSnapshot();
  });

  it("should show model info", () => {
    const title = gabTitle('模型信息')[0];
    act(() => {
      title.click();
    });
    expect(gContainer).toMatchSnapshot();
  });

  it("should refresh model info", () => {
    const title = gabTitle('模型信息')[0];
    act(() => {
      title.click();
    });
    const modelManager = store.getState().system.viewer3D.getViewerImpl().modelManager;
    const tmpInfo = modelManager.getStatisticsInfo;
    modelManager.getStatisticsInfo = () => ({
      componentCount: 100,
      trianglesCount: 200,
    });
    act(() => {
      title.click();
    });
    act(() => {
      title.click();
    });
    expect(gContainer).toMatchSnapshot();
    // 测试完毕后复原store
    modelManager.getStatisticsInfo = tmpInfo;
  });
});
