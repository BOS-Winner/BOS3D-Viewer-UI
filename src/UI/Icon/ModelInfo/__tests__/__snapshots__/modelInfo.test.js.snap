// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modelInfo should refresh model info 1`] = `
<div>
  <div
    role="button"
    tabindex="0"
    title="模型信息"
  >
    <div
      class="icon "
      data-selected="true"
      role="button"
      style="background-image: url(modelInfo.png);"
      tabindex="0"
    />
    <div
      style="display: block; position: absolute; top: 40%; width: 300px; height: 140px;"
      title="模型信息"
    >
      <div>
        <span>
          模型信息
        </span>
        <button
          data-testid="close-mock-modal"
          type="button"
        >
          close mock modal
        </button>
      </div>
      <div
        class="modelInfo"
      >
        <div
          class="item"
        >
          <div
            class="name"
          >
            三角面数：
          </div>
          200
        </div>
        <div
          class="item"
        >
          <div
            class="name"
          >
            构件数：
          </div>
          100
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`modelInfo should render dom 1`] = `
<div>
  <div
    role="button"
    tabindex="0"
    title="模型信息"
  >
    <div
      class="icon "
      data-selected="false"
      role="button"
      style="background-image: url(modelInfo.png);"
      tabindex="0"
    />
    <div
      style="display: none; position: absolute; top: 40%; width: 300px; height: 140px;"
      title="模型信息"
    >
      <div>
        <span>
          模型信息
        </span>
        <button
          data-testid="close-mock-modal"
          type="button"
        >
          close mock modal
        </button>
      </div>
      <div
        class="modelInfo"
      >
        <div
          class="item"
        >
          <div
            class="name"
          >
            三角面数：
          </div>
          0
        </div>
        <div
          class="item"
        >
          <div
            class="name"
          >
            构件数：
          </div>
          0
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`modelInfo should show model info 1`] = `
<div>
  <div
    role="button"
    tabindex="0"
    title="模型信息"
  >
    <div
      class="icon "
      data-selected="true"
      role="button"
      style="background-image: url(modelInfo.png);"
      tabindex="0"
    />
    <div
      style="display: block; position: absolute; top: 40%; width: 300px; height: 140px;"
      title="模型信息"
    >
      <div>
        <span>
          模型信息
        </span>
        <button
          data-testid="close-mock-modal"
          type="button"
        >
          close mock modal
        </button>
      </div>
      <div
        class="modelInfo"
      >
        <div
          class="item"
        >
          <div
            class="name"
          >
            三角面数：
          </div>
          1919810
        </div>
        <div
          class="item"
        >
          <div
            class="name"
          >
            构件数：
          </div>
          114514
        </div>
      </div>
    </div>
  </div>
</div>
`;
